mesa/visualization/components/matplotlib.py

```diff
@@ -146,132 +180,269 @@ def draw_property_layers(ax, space, propertylayer_portrayal, model):
-            c=[c for c, show in zip(portray_data["c"], mask) if show],
-            marker=marker,
+    """
+    arguments = {"loc": [], "s": [], "c": [], "marker": []}
+    for agent in space.agents:
+        portray = agent_portrayal(agent)
+        loc = agent.pos
+        if loc is None:
+            loc = agent.cell.coordinate
+
+        arguments["loc"].append(loc)
+        arguments["s"].append(portray.get("s", s_default))
+        arguments["c"].append(portray.get("c", c_default))
+        arguments["marker"].append(portray.get("marker", marker_default))
+
+    return {k: np.asarray(v) for k, v in arguments.items()}
+
+
+def draw_orthogonal_grid(
+    space: OrthogonalGrid, agent_portrayal: Callable, ax, draw_grid: bool = True
+):
+    """Visualize a orthogonal grid.
+
+    Args:
+        space: the space to visualize
+        agent_portrayal: a callable that is called with the agent and returns a dict
+        ax: a Matplotlib Axes instance
+        draw_grid: whether to draw the grid
+
+    Returns:
+        A Figure and Axes instance
+
+    """
+    # gather agent data
+    s_default = (180 / max(space.width, space.height)) ** 2
+    arguments = collect_agent_data(space, agent_portrayal, s_default=s_default)
+
+    # plot the agents
+    _scatter(ax, arguments)
+
+    # further styling
+    ax.set_xlim(-0.5, space.width - 0.5)
+    ax.set_ylim(-0.5, space.height - 0.5)
+
+    if draw_grid:
+        # Draw grid lines
+        for x in np.arange(-0.5, space.width - 0.5, 1):
+            ax.axvline(x, color="gray", linestyle=":")
+        for y in np.arange(-0.5, space.height - 0.5, 1):
+            ax.axhline(y, color="gray", linestyle=":")
+
+
+def draw_hex_grid(
+    space: HexGrid, agent_portrayal: Callable, ax, draw_grid: bool = True
+):
+    """Visualize a hex grid.
+
+    Args:
+        space: the space to visualize
+        agent_portrayal: a callable that is called with the agent and returns a dict
+        ax: a Matplotlib Axes instance
+        draw_grid: whether to draw the grid
+
+    Returns:
+        A Figure and Axes instance
+
+    """
+    # gather data
+    s_default = (180 / max(space.width, space.height)) ** 2
+    arguments = collect_agent_data(space, agent_portrayal, s_default=s_default)
+
+    # for hexgrids we have to go from logical coordinates to visual coordinates
+    # this is a bit messy.
+
+    # give all even rows an offset in the x direction
+    # give all rows an offset in the y direction
+
+    # numbers here are based on a distance of 1 between centers of hexes
+    offset = math.sqrt(0.75)
+
+    loc = arguments["loc"].astype(float)
+
+    logical = np.mod(loc[:, 1], 2) == 0
+    loc[:, 0][logical] += 0.5
+    loc[:, 1] *= offset
+    arguments["loc"] = loc
+
+    # plot the agents
+    _scatter(ax, arguments)
+
+    # further styling and adding of grid
+    ax.set_xlim(-1, space.width + 0.5)
+    ax.set_ylim(-offset, space.height * offset)
+
+    def setup_hexmesh(
```

Comment:
No idea how it works, but can we maybe use matplotlibs hexbin? https://matplotlib.org/stable/plot_types/stats/hexbin.html#sphx-glr-plot-types-stats-hexbin-py

Reply:
Looking at the source of hexbin, it uses a polycollection instead of patchcollection as I do here. I am fine with taking a closer look to see if polycollection creates a cleaner grid. I need to look at that anyway also for plotting propertylayers. 

However, I would prefer to leave that for a future PR.